### ISSUE TITLE
TQDM progress bars in jupyter notebooks do not appear in PDF via latex output 

### ISSUE THREAD

**Issue by EdwardRaff** [tqdm/tqdm#937]
- [X] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [X] visual output bug
    + [ ] documentation request (i.e. "X is missing from the documentation." If instead I want to ask "how to use X?" I understand [StackOverflow#tqdm] is more appropriate)
    + [ ] new feature request
- [X] I have visited the [source website], and in particular
  read the [known issues]
- [X] I have searched through the [issue tracker] for duplicates
- [X] I have mentioned version numbers, operating system and
  environment, where applicable:
  ```python
  import tqdm, sys
  print(tqdm.__version__, sys.version, sys.platform)
  ```
gives:
```
4.31.1 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0] linux
```

When using the jupyter notebook extensions of TQDM, we get nice beautiful progress bars in the notebook, which is awesome. I'm currently writing some lectures for my students, and want to share a PDF version as well. But when I export to PDF via Latex, the progress bars are replaced by raw commands in the background like `HBox` (see attachment for an example)
[tqdm_pdf_example.pdf](https://github.com/tqdm/tqdm/files/4466741/tqdm_pdf_example.pdf)

It would be awesome if there was a way to get tqdm progress bars to show up in the latex output. I'm not sure if this is a tqdm or jupyter issue though. 

  [source website]: https://github.com/tqdm/tqdm/
  [known issues]: https://github.com/tqdm/tqdm/#faq-and-known-issues
  [issue tracker]: https://github.com/tqdm/tqdm/issues?q=
  [StackOverflow#tqdm]: https://stackoverflow.com/questions/tagged/tqdm

**Comment by casperdcl** [comment]
I'm not sure there's a clean way to do this. I haven't been able to follow the documentation for `ipywidgets`; but managed to get a hack working:

```python
from tqdm.notebook import tqdm
from IPython import display as disp
from ipywidgets import HBox

bar = tqdm(total=10, initial=2)

# hack: display in such a way that export to plain text works
class TqdmHBox(HBox):
    def __repr__(self):
        return bar.format_meter(**dict(
            bar.format_dict,
            bar_format=bar.format_dict['bar_format'].replace("<bar/>", "{bar}"),
            ascii=True))
box = TqdmHBox(children=bar.container.children)
disp.clear_output(True)
disp.display(box)
```

Could you try this and confirm if it works? When exporting to LaTeX, you should get something like:

```
 20%|##        | 2/10 [00:00<00:00, 40.52it/s]
```

If so, I can put this into `notebook`.

Ideally some `ipywidgets` dev should confirm if this is the best way to override display of `HBox` in non-HTML envs...

**Comment by EdwardRaff** [comment]
I get the below error with your code:

```
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-2-1ff07c714c7d> in <module>
----> 1 from tqdm.notebook import tqdm
      2 from IPython import display as disp
      3 from ipywidgets import HBox
      4 
      5 bar = tqdm(total=10, initial=2)

ModuleNotFoundError: No module named 'tqdm.notebook'
```
I was using `from tqdm.autonotebook import tqdm` before, but if I do that your code gives me this new error 

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-3ef21aa8464f> in <module>
     12             bar_format=bar.format_dict['bar_format'].replace("<bar/>", "{bar}"),
     13             ascii=True))
---> 14 box = TqdmHBox(children=bar.container.children)
     15 disp.clear_output(True)
     16 disp.display(box)

AttributeError: 'tqdm_notebook' object has no attribute 'container'
```

**Comment by casperdcl** [comment]
can you update to the latest `tqdm` version and try again?

**Comment by EdwardRaff** [comment]
Updated to latest and it does work! An ASCII bar shows up in the PDF. Obviously not as nice as the regular TQDM bar, but _much_ better!

**Comment by casperdcl** [comment]
yes I don't think I can improve on an ascii bar - having a nicer looking png would definitely be upstream https://github.com/jupyter-widgets/ipywidgets/issues/2280.

Note that even when saving, closing & re-opening a notebook I get:

```
SEJveChjaGlsZHJlbj0oRmxvYXRQcm9ncmVzcyh2YWx1ZT0wLjAsIGRlc2NyaXB0aW9uPXUnTEEtR0FOJywgbWF4PTIwMC4wLCBzdHlsZT1Qcm9ncmVzc1N0eWxlKGRlc2NyaXB0aW9uX3dpZHTigKY=
HBox(children=(FloatProgress(value=0.0, max=3.0), HTML(value=u'')))
```

This seems like a different bug also with upstream. Surely a notebook should be able to save and reload its own HTML?

Anyway after my LaTeX fix the notebook case will also show an ascii bar upon reload.

**Comment by casperdcl** [comment]
sigh... the `__repr__` thing doesn't work. Simple example:

```python
from ipywidgets import Output
from IPython import display as disp

class MyOutput(Output):
    def __init__(self):
        super(Output, self).__init__()
        self.last_display = "Nothing to display yet"
    def __repr__(self):
        return repr(self.last_display)
out = MyOutput()
disp.display(out)

with out:
    disp.display("Some text!")
```

This correctly displays `'Some text!'` in the notebook, but when exported shows `'Nothing to display yet'`. If `__repr__` is not overridden, exported versions just show `Output()`.

**Comment by EdwardRaff** [comment]
Hi, i notice you did a new commit addressing this issue. Does this mean the head branch / next release of TQDM should automatically output the ASCII bar? 

If so I'm happy to test against head branch and close the issue if it works again!

**Comment by casperdcl** [comment]
No it's not merged yet (i.e. it's not yet done automatically) because I suspect there's an upstream issue preventing automation (https://github.com/jupyter-widgets/ipywidgets/issues/2850).

For now all I can suggest is tidying up & using my original hack.

**Comment by EdwardRaff** [comment]
Is there a way to automatically apply your "hack" to every use of TQDM?

**Comment by casperdcl** [comment]
```python
from tqdm.notebook import tqdm_notebook

class tqdm(tqdm_notebook):
    def __init__(self, *args, **kwargs):
        super(tqdm, self).__init__(*args, **kwargs)
        from IPython import display as disp
        from ipywidgets import HBox

        class TqdmHBox(HBox):
            """hack: display in such a way that export to plain text works"""
            def __repr__(__self):
                return self.format_meter(**dict(
                    self.format_dict,
                    bar_format=self.format_dict['bar_format'].replace("<bar/>", "{bar}"),
                    ascii=True))

        self.container = box = TqdmHBox(children=self.container.children)
        disp.clear_output(True)
        disp.display(box)
```

you can put that into one `somefile.py` and then do `from somefile import tqdm` in all your notebooks.

**Comment by casperdcl** [comment]
slightly better version:

```python
from tqdm.notebook import tqdm_notebook

class tqdm_n(tqdm_notebook):
    def __init__(self, *args, **kwargs):
        super(tqdm_n, self).__init__(*args, **kwargs)
        try:
            self.container.close()
        except AttributeError:
            self.container.visible = False
        from IPython import display as disp
        from ipywidgets import HBox

        class TqdmHBox(HBox):
            """hack: display in such a way that export to plain text works"""
            def __repr__(__self):
                return self.format_meter(**dict(
                    self.format_dict,
                    bar_format=self.format_dict['bar_format'].replace("<bar/>", "{bar}"),
                    ascii=True))
        self.container = box = TqdmHBox(children=self.container.children)
        disp.display(box)
```

**Comment by casperdcl** [comment]
@EdwardRaff can you try #1099 (`pip install -U "git+https://github.com/tqdm/tqdm.git@5356146#egg=tqdm"`)? No "hacks" should be required... If it works, I can release the fix in `tqdm>=4.55.0`
